src/groupByKey.ts

```diff
@@ -2,6 +2,29 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+export function groupByKey<ObjectToSort, Key extends keyof ObjectToSort>(
+  items: ObjectToSort[],
+  key: Key,
+): GroupsMap<ObjectToSort> {
+  if (items.length === 0) {
+    return {};
+  }
+
+  type ValueFromObject = ObjectToSort[Key];
+
+  const valuesOfKey: Array<ValueFromObject> = items.map(
+    (item: ObjectToSort) => item[key],
+  );
+  const uniqueValuesOfKey: Set<ValueFromObject> = new Set([...valuesOfKey]);
+  const sortedObjects: GroupsMap<ObjectToSort> = {};
+
+  uniqueValuesOfKey.forEach((uniqueValueOfKey: ValueFromObject) => {
```

Comment:
The filter operation inside the forEach loop results in a quadratic time complexity. This can be avoided by using a single reduce function to create the grouped object.